subsys/net/ip/tcp.c

```diff
@@ -2639,124 +2671,205 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
+			 * of the packet to check the sequence number of the FIN flag with the ACK
+			 */
+			if (FL(&fl, &, FIN, (th_seq(th) + len) == conn->ack)) {
 				conn_ack(conn, + 1);
-				tcp_out(conn, ACK);
-				next = TCP_TIME_WAIT;
-				verdict = NET_OK;
-			} else if (len > 0) {
-				/* Send out a duplicate ACK */
-				tcp_out(conn,  ACK);
+				NET_DBG("FIN received, going to TIME WAIT");
+
+				next = tcp_enter_time_wait(conn);
+
 				verdict = NET_OK;
+				tcp_out(conn, ACK);
+			} else {
+				if (len > 0) {
+					/* Send out a duplicate ACK */
+					tcp_out(conn, ACK);
+					verdict = NET_OK;
+				}
 			}
 		}
 		break;
 	case TCP_CLOSING:
-		if (th && FL(&fl, ==, ACK, th_seq(th) == conn->ack)) {
-			tcp_send_timer_cancel(conn);
-			next = TCP_TIME_WAIT;
-			verdict = NET_OK;
+		if (th) {
+			bool fin_acked = false;
+
+			/*
+			 * Closing:
+			 * Our FIN has to be acknowledged
+			 * -   -> TCP_CLOSING
+			 * ACK -> TCP_TIME_WAIT
+			 */
+			int32_t new_len = tcp_compute_new_length(conn, th, len, true);
+
+			if (new_len > 0) {
+				/* This should not happen here, as no data can be send after
+				 * the FIN flag has been send.
+				 */
+				NET_ERR("conn: %p, new bytes %u during CLOSING state "
+					"sending reset", conn, new_len);
+				net_stats_update_tcp_seg_drop(conn->iface);
+
+				next = tcp_enter_time_wait(conn);
+
+				tcp_out(conn, RST);
+				break;
+			}
+
+			if (FL(&fl, &, ACK, th_ack(th) == conn->seq)) {
+				NET_DBG("FIN acknowledged, going to TIME WAIT "
+					"state seq %u, ack %u"
+					, conn->seq, conn->ack);
+
+				next = tcp_enter_time_wait(conn);
+				fin_acked = true;
+
+				verdict = NET_OK;
+			}
+
+			/*
+			 * There can also be data in the message, so compute with the length
+			 * of the packet to check with the ack
+			 */
+			if ((FL(&fl, &, FIN, (th_seq(th) + len) == conn->ack)) || (len > 0)) {
+				if (fin_acked) {
+					/* Send out a duplicate ACK */
+					tcp_out(conn, ACK);
+				} else {
+					/* Send out a duplicate ACK, with the pending FIN
+					 * flag
+					 */
+					tcp_send_timer_cancel(conn);
+					tcp_out_ext(conn, FIN | ACK, NULL, conn->seq - 1);
+				}
+				verdict = NET_OK;
+			}
 		}
 		break;
 	case TCP_TIME_WAIT:
-		/* Acknowledge any FIN attempts, in case retransmission took
-		 * place.
-		 */
-		if (th && (FL(&fl, ==, (FIN | ACK), th_seq(th) + 1 == conn->ack) ||
-			   FL(&fl, ==, FIN, th_seq(th) + 1 == conn->ack))) {
-			tcp_out(conn, ACK);
-			verdict = NET_OK;
-		}
+		if (th) {
+			int32_t new_len = tcp_compute_new_length(conn, th, len, true);
 
-		k_work_reschedule_for_queue(
-			&tcp_work_q, &conn->timewait_timer,
-			K_MSEC(CONFIG_NET_TCP_TIME_WAIT_DELAY));
+			if (new_len > 0) {
+				/* This should not happen here, as no data can be send after
+				 * the FIN flag has been send.
+				 */
+				NET_ERR("conn: %p, new bytes %u during TIME-WAIT state "
+					"sending reset", conn, new_len);
+				net_stats_update_tcp_seg_drop(conn->iface);
+
+				tcp_out(conn, RST);
```

Comment:
And here.

Reply:
To enter the TCP_TIME_WAIT state, tcp_send_timer_cancel, must have been called before, so I do not think it is required here. The only way to enter TIME_WAIT is through tcp_enter_time_wait, which calls tcp_send_timer_cancel. Packets with RST, ACK, PSH or ACK | PSH do not trigger the retransmit timer,